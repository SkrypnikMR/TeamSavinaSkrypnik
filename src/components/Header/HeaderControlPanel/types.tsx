import React from 'react';

export interface IDiv {
    display: string;
    justifyContent: string;
    alignItems: string;
    height: string;
    width: string;
    color:string;
    marginRight: string;
    marginLeft: string;
    cursor: string;
}
