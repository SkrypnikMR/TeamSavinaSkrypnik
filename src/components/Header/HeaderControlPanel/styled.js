import styled from 'styled-components';

export const StControl = styled.div`
        display: flex;
        justify-content: flex-end;
        align-items: center;
        height: 100%;
        width: 70%;
        p{
                color:white;
        }
`;
