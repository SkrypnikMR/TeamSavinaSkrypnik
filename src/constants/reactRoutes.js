export const APP_ROUTES = {
    registration: '/registration',
    login: '/',
    chat: '/chat',
    account: '/account',
};
