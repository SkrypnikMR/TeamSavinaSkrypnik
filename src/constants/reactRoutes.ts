interface IRoutes {
    registration: string;
    login: string;
}

export const APP_ROUTES: IRoutes = {
    registration: '/registration',
    login: '/',
};
