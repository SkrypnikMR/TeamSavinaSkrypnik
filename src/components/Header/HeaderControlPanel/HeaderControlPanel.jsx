import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import { StControl } from './styled';
import { HEADER_CONTROL_BTNS } from '/src/constants/componentsÐ¡onsts';
import Button from '../../UI/Button';
import { APP_ROUTES } from '/src/constants/reactRoutes';
import { ROUTS_WITHOUT_MY_ACCOUNT } from '../../../constants/ui';
import { support } from '../../../helpers/support';
import { colorDefault } from '../../UI/baseLayout';

const HeaderControlPanel = ({
    themeMode,
    setValue,
    history,
    location,
    logOut,
    userNotifSettings,
    onlineUsersCount }) => {
    const { i18n } = useTranslation();
    const handleChangeLanguage = (e) => {
        i18n.changeLanguage(e.target.value);
        localStorage.setItem('lang', e.target.value);
    };
    const handleThemeClick = ({ target }) => {
        support.setSessionStorageItem('themeMode', target.value);
        setValue({ name: 'themeMode', value: target.value });
    };
    const handleLogOutClick = () => {
        logOut();
        history.push(APP_ROUTES.login);
    };
    const handleNotifClick = (e) => {
        setValue({ name: 'settings', value: { notifications: Boolean(e.target.value) } });
        support.setSessionStorageItem('settings', { notifications: Boolean(e.target.value) });
    };
    const handleMyAccountClick = () => history.push(APP_ROUTES.account);
    const getFunctionForButtons = (el) => {
        switch (el.id) {
            case 'theme_btn': return handleThemeClick;
            case 'logOut': return handleLogOutClick;
            case 'account': return handleMyAccountClick;
            case 'notif_btn': return handleNotifClick;
            default: return handleChangeLanguage;
        }
    };
    return (
        <StControl >
            <p>
                {location.pathname === '/chat'
                    ? `${i18n.t('online')}: ${onlineUsersCount === 0 ? 0 : onlineUsersCount - 1}`
                    : null}
            </p>
            {' '}
            {HEADER_CONTROL_BTNS.map((el) => {
                if (el.value === themeMode) return null;
                if (el.id === 'notif_btn' && Boolean(el.value) === userNotifSettings) return null;
                if ((el.rout === '/account' && ROUTS_WITHOUT_MY_ACCOUNT.includes(location.pathname))
                    || el.rout === location.pathname) return null;
                return (
                    <Button
                        id={el.id}
                        content={el.content}
                        key={el.id}
                        color={colorDefault}
                        fontSize='26px'
                        width='60px'
                        height="10vh"
                        borderRadius="0px"
                        value={el.value}
                        bgColor="transparent"
                        onClick={getFunctionForButtons(el)}
                    />
                );
            })}
        </StControl>
    );
};

HeaderControlPanel.propTypes = {
    setValue: PropTypes.func,
    themeMode: PropTypes.string,
    history: PropTypes.object,
    location: PropTypes.object,
    logOut: PropTypes.func,
    userNotifSettings: PropTypes.bool,
    onlineUsersCount: PropTypes.number,
};

export default HeaderControlPanel;
