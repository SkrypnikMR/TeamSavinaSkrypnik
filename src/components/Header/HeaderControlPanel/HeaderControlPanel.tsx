import React from 'react';
import { useTranslation } from 'react-i18next';
import { APP_ROUTES } from 'src/constants/reactRoutes';

import { HEADER_CONTROL_BTNS } from 'src/constants/componentsÐ¡onsts';
import Button from '../../UI/Button';
import { support } from '../../../helpers/support';
import { colorDefault } from '../../UI/baseLayout';
import { useTheme } from 'src/components/Hook/useTheme';

import { StControl } from './styled';

const HeaderControlPanel = ({
    history,
    location }) => {
    const { i18n } = useTranslation();
    const { changeTheme } = useTheme();
    const handleChangeLanguage = (e) => {
        i18n.changeLanguage(e.target.value);
        localStorage.setItem('lang', e.target.value);
    };
    const handleThemeClick = ({ target }) => {
        changeTheme();
        // support.setSessionStorageItem('themeMode', target.value);
    };
    const handleLogOutClick = () => {
        history.push(APP_ROUTES.login);
    };
    const handleNotifClick = (e) => {
        support.setSessionStorageItem('settings', { notifications: Boolean(e.target.value) });
    };
    const handleMyAccountClick = () => history.push(APP_ROUTES.account);
    const getFunctionForButtons = (el) => {
        switch (el.id) {
            case 'theme_btn': return handleThemeClick;
            case 'logOut': return handleLogOutClick;
            case 'account': return handleMyAccountClick;
            case 'notif_btn': return handleNotifClick;
            default: return handleChangeLanguage;
        }
    };
    return (
        <StControl >
            {' '}
            {HEADER_CONTROL_BTNS.map((el) => {
                return (
                    <Button
                        id={el.id}
                        content={el.content}
                        key={el.id}
                        color={colorDefault}
                        fontSize='26px'
                        width='40px'
                        borderRadius="0px"
                        value={el.value}
                        bgColor="transparent"
                        onClick={getFunctionForButtons(el)}
                    />
                );
            })}
            <img src="../../../../public/assets/images/log-out.png"/>
        </StControl>
    );
};

export default HeaderControlPanel;
