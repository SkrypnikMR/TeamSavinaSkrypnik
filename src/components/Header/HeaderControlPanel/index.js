import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { setValue } from 'src/Store/user/actions';
import { userThemeMode, userNotifSettings } from 'src/Store/user/selectors';
import HeaderControlPanel from './HeaderControlPanel';

const mapStateToProps = state => ({
    themeMode: userThemeMode(state),
    userNotifSettings: userNotifSettings(state),
});
const mapDispatchToProps = dispatch => ({
    setValue: payload => dispatch(setValue(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(HeaderControlPanel));
