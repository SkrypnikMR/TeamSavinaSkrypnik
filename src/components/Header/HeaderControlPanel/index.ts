import { withRouter } from 'react-router';
import HeaderControlPanel from './HeaderControlPanel';

export default withRouter(HeaderControlPanel);
