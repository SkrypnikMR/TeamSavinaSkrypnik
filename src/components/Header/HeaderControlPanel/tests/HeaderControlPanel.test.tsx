import React from 'react';
import { shallow, mount } from 'enzyme';
import { useTranslation } from 'react-i18next';
import configureStore from 'redux-mock-store';
import HeaderControlPanel from '../HeaderControlPanel';
import { mountSmart } from '../../../../../__tests__/testHelper';
import { APP_ROUTES } from '../../../../constants/reactRoutes';

jest.mock('react-i18next', () => ({ useTranslation: jest.fn().mockReturnValue({ i18n: { changeLanguage: jest.fn() }, t: jest.fn() }) }));

const mockStore = configureStore();
const store = mockStore({
  game: {
     rooms: [],
    userLogin: '',
    actualRoom: {
      gameType: '',
      creatorLogin: '',
      guestLogin: '',
      startTime: 0,
      id: '',
      stepDoList: [],
    },
    stepOrder: '',
    stepHistory: [],
    winner: '',
    possibleSteps: [],
  },
});

describe('HeaderControlPanel', () => {
    let props: any;
    let portal: any;
    const realLocalStorageClear = localStorage.clear;
    beforeEach(() => {
        portal = document.createElement('div');
        portal.setAttribute('id', 'portal');
        document.body.append(portal);
        props = { history: { push: jest.fn() } };
        localStorage.clear = jest.fn();
    });
    afterEach(() => {
        localStorage.clear = realLocalStorageClear;
        document.body.removeChild(portal);
    });

    it('Should match snapshot', () => {
        const component = shallow(<HeaderControlPanel {...props} />);
        expect(component.html()).toMatchSnapshot();
    });
    it('Should render 3button and 1 img', () => {
        const component = mount(<HeaderControlPanel {...props} />);
        expect(component.find('button')).toHaveLength(3);
        expect(component.find('img')).toHaveLength(1);
    });
    it('Should call useTheme', () => {
        const component = mount(<HeaderControlPanel {...props} />);
        component.find('button').at(0).simulate('click');
    });
    it('Should call i18n.changeLanguage with en', () => {
        const { i18n } = useTranslation();
        const component = mount(<HeaderControlPanel {...props} />);
        component.find('button').at(1).simulate('click');
        expect(i18n.changeLanguage).toHaveBeenCalledWith('en');
    });
    it('Should call i18n.changeLanguage with ru', () => {
        const { i18n } = useTranslation();
        const component = mount(<HeaderControlPanel {...props} />);
        component.find('button').at(1).simulate('click');
        expect(i18n.changeLanguage).toHaveBeenCalledWith('en');
    });
    it('Should call history.push and localStorage.clear', () => {
        const component = mount(<HeaderControlPanel {...props} />);
        component.find('img').simulate('click');
        expect(props.history.push).toHaveBeenCalledWith(APP_ROUTES.login);
    });
    it('Should render Modal', () => {
        props.actualRoomId = 'yasdyasdhsaidsaod';
        const component = mountSmart(<HeaderControlPanel {...props} />, store);
        component.find('img').simulate('click');
        expect(component.find('ModalCustom')).toHaveLength(1);
    });
});
