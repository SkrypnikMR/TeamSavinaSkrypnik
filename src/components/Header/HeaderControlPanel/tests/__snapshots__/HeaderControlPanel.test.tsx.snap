// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderControlPanel Should match snapshot 1`] = `"<div class=\\"styled__StControl-sc-1l7fu2q-0 bcijKV\\"> <button id=\\"theme_btn\\" type=\\"button\\" value=\\"dark\\" color=\\"white\\" font-size=\\"26px\\" width=\\"40px\\" class=\\"styled__StButton-sc-1y35zax-0 kQlZMx\\"></button><button id=\\"en\\" type=\\"button\\" value=\\"en\\" color=\\"white\\" font-size=\\"26px\\" width=\\"40px\\" class=\\"styled__StButton-sc-1y35zax-0 kLPgkh\\"></button><button id=\\"ru\\" type=\\"button\\" value=\\"ru\\" color=\\"white\\" font-size=\\"26px\\" width=\\"40px\\" class=\\"styled__StButton-sc-1y35zax-0 iwWvRJ\\"></button><img src=\\"public/assets/images/log-out.png\\"/></div>"`;
